### PR TITLE
[RISCV][dev] Platform-Level Interrupt Controller (PLIC) driver.

### PR DESCRIPTION
**Pull request by MichalBlk**
This change introduces a device driver for the Platform-Level Interrupt Controller device.

This PR contains changes from #1235, #1251.

### CODE REVIEW

**Review by cahirwpz — APPROVED** [review]

File: sys/drv/plic.c [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+
+  uint32_t en = in4(PLIC_ENABLE_SV(irq));
+  en |= 1 << (irq % 32);
+  out4(PLIC_ENABLE_SV(irq), en);
+}
+
+static const char *plic_intr_name(unsigned irq) {
+  char buf[32];
+  snprintf(buf, sizeof(buf), "PLIC source %u", irq);
+  return kstrndup(M_STR, buf, sizeof(buf));
```
Comment (cahirwpz) [inline sys/drv/plic.c:64]: Is there a chance you will add `kasprintf` function to the kernel?

File: sys/drv/plic.c [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+
+static int plic_map_intr(device_t *pic, device_t *dev, phandle_t *intr,
+                         int icells) {
+  plic_state_t *plic = pic->state;
+
+  if (icells != 1)
+    return -1;
+
+  unsigned irq = *intr;
+  if (!irq || irq > plic->ndev)
```
Comment (cahirwpz) [inline sys/drv/plic.c:106]: Of by one mistake? Shouldn't it be `irq >= plic->ndev` ?

File: sys/drv/plic.c [hunk 3]
```diff
@@ -0,0 +1,190 @@
+#define KL_LOG KL_DEV
+#include <sys/bus.h>
+#include <sys/devclass.h>
+#include <sys/errno.h>
+#include <sys/fdt.h>
+#include <sys/interrupt.h>
+#include <sys/klog.h>
+#include <sys/libkern.h>
+
+/* PLIC memory map. */
```
Comment (cahirwpz) [inline sys/drv/plic.c:10]: Could you put a link to the specification here?

File: sys/drv/plic.c [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+  return irq;
+}
+
+static intr_filter_t plic_intr_handler(void *arg) {
+  plic_state_t *plic = arg;
+
+  /* Claim any pending interrupt. */
+  uint32_t irq = in4(PLIC_CLAIM_SV);
+
+  if (irq) {
```
Comment (cahirwpz) [inline sys/drv/plic.c:118]: I prefer if-clause to filter out weird cases leaving the main task of function to be implemented at the very bottom.

File: sys/drv/plic.c (hunk 2 above)
Comment (MichalBlk) [reply sys/drv/plic.c:106]: No. It's fine (See https://mimiker.ii.uni.wroc.pl/source/xref/FreeBSD/sys/riscv/riscv/plic.c?r=80f39bd9#219).

File: sys/drv/plic.c (hunk 4 above)
Comment (MichalBlk) [reply sys/drv/plic.c]: PLIC spec: "An interrupt ID of 0 is reserved to mean “no interrupt”."
I changed the control flow and inserted a panic on this faulty condition.

File: sys/drv/plic.c (hunk 3 above)
Comment (MichalBlk) [reply sys/drv/plic.c:16]: Done.

File: sys/drv/plic.c (hunk 1 above)
Comment (MichalBlk) [reply sys/drv/plic.c:70]: #1251. Done.

**Pull request merged**